L8-c++-internalization.md

```diff
@@ -0,0 +1,172 @@
+Separate internal-only and public sections of the C++ wrapping
+----
+* Author(s): vjpai
+* Approver: ctiller
+* Status: Approved
+* Implemented in: C++
+* Last updated: July 6, 2017
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/Z8S9P4NzGHs
+
+## Abstract
+
+Maintain separation of C++ classes and functions that are intended
+to be directly used by application developers from those that are
+supposed to be used only through codegen or for that exist for gRPC's
+internal implementation (e.g., interactions with the core API).
+
+## Background
+
+Essentially all of gRPC C++ lives in `namespace grpc`, giving the
+appearance that all portions are equally intended as part of the
+publicly-exposed API. This gRFC aims to clarify this by categorizing
+the gRPC C++ wrapping, moving some components to `namespace
+grpc::internal`, and moving some class constructors to `private` (with
+friends or appropriately-named `static` factories for internalized generation).
+
+### Related Proposals: 
+
+N/A
+
+## Proposal
+
+The C++ gRPC wrapping includes numerous classes and functions, but
+they fall into five separate categories:
+
+1. Documented for public use
+1. Intended for interfacing with serialization layers such as protobuf
+1. Intended for use through the code-generation layer
```

Comment:
The first one should probably also say "Documented for public use by end-user applications" just to make it clear that we're talking about client and server applications in that category and not other libraries like flatbuffers.